### ISSUE TITLE
Reactotron not connecting

### ISSUE THREAD

**Issue by JonCognioDigital** [infinitered/reactotron#815]
Hi,

I installed Reactotron tonight and it seemed to work well at first, I started installing plugins and playing with logging. It was all going great until all of a sudden it won't connect.

No errors. just nothing....

I am using Expo, a real Android device using Expo's tunneling, my lan IP is 192.168.0.22:19000

Here is my config

```
import url from "url"
import { NativeModules } from "react-native";
import Reactotron, { asyncStorage, openInEditor, networking, trackGlobalErrors } from 'reactotron-react-native'
import { reactotronRedux } from "reactotron-redux"
import apisaucePlugin from 'reactotron-apisauce'

const { hostname } = url.parse(NativeModules.SourceCode.scriptURL)

const reactotron = Reactotron
  .configure({host: hostname })
  .use(networking())
  .use(reactotronRedux())
  .use(asyncStorage())
  .use(openInEditor())
  .use(trackGlobalErrors())
  .use(apisaucePlugin({
    // ignoreContentTypes: /^(image)\/.*$/i   // <--- a way to skip printing the body of some requests (default is any image)
  }))
  .useReactNative()
  .connect()

  export default reactotron
```

**Comment by rmevans9** [comment]
He removed them as they are setup via the `.useReactNative()` bit. `useReactNative` is a helper call to setup all plugins that are shipped with `reactotron-react-native`.

ref: https://github.com/infinitered/reactotron/blob/master/packages/reactotron-react-native/src/index.js#L70

**Comment by jon64digital** [comment]
Thanks,

I can see from that page that the only plugin not ncluded by default is apiSauce so I'll keep that in.

Unfortunately, tonight it has stopped working again completely. I had to switch to using the expo app because for some reason Epo had stopped showing console.logs in the console (or anything after starting the package manager actually which was some kind of bug I couldn't work out). Since switching to the Expo app Reactotron hasn't worked, despite still being in LAN mode on the same API and desite moving the .connect() call to App.js componentDidMount (which had previously fixed the problem). 

I'm now just getting absolutely nothing from Reactotron, but it is still connecting to React Native Debugger with no problems.

**Comment by jon64digital** [comment]
I ran netstat to see if anything was listening on port 9090 and it seems like every few secons Reactotron is trying to establish a connection from a different port on an internal IP ending in 26. Is this right?

[Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37266     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37274     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37283     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37291     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37299     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37306     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37313     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37324     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37339     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37358     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37367     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37375     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37385     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37398     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37407     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37425     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37432     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37440     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:37477     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:49936     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:49943     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:49956     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:49964     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:49969     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:49972     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:49980     ESTABLISHED
 [Reactotron.exe]
  TCP    192.168.0.22:9090      192.168.0.26:49987     ESTABLISHED
 [Reactotron.exe]

**Comment by rmevans9** [comment]
I know Expo and reactotron has had some connectivity issues. In your configure I would suggest adding your IP as the host to see if that makes it work. Something like this:

`.configure({ host: '<YOUR IP>' })`

Yes, `apisauce` plugin is not on there as it is not shipped with `reactotron-react-native`. That said the `networking` plugin supersedes the `apisauce` plugin as it listens to all network traffic in RN (at least in the RN side, not the native side)

**Comment by jon64digital** [comment]
Ah ok, I was wondering what else it would give me but if it doesn't then I'll remove it. I am already configuring my hostName (you can see how in my first post) but it doesn't seem to make any difference now. The Reactotron client doesn't seem to have any options so I'm running out of ideas trying to work out if my code isn't broadcasting correctly or Reactotron isn't listening correctly. There must be a mismatch somewhere but hunting it down is difficult.

**Comment by jon64digital** [comment]
OK, so I think I'm ready to give up on this now and just use React Native Debugger. Which is a shame as it looks like Reactotron has some nice features. Thanks for trying to help me out but I'm now back to the most basic implementation, no plugins, no redux, just a simple .configure() with the correct hostname and connect() inside my App's componentDidMount and unfortunately the Reactotron client is completely unresponsive and sees no connections.

**Comment by bsendras** [comment]
@jon64digital I had the same problem with the project that I am currently working on. Reactotron.app never received a connection from my application. No errors, no logs, no nothing. I tested it on a separate brand new app, following the minimal config as is explained in the [RADME](https://github.com/infinitered/reactotron) file. That worked fine. But there were no difference between both apps configs. So I check both package.json files.

What my non-working project had:
```json
    "reactotron-react-native": "^1.14.0",
    "reactotron-redux": "^1.13.0", // because I keep the navigation state in my redux store, you might not have/need this.
```
What my new test working project had:
```
    "reactotron-react-native": "^2.1.0",
    "reactotron-redux": "^2.1.0",
```
So I installed those new versions of reactotron dependencies, build the app again and finally it was able to connect to Reactotron.app(v2.1.2 by the way).

Hope, it helps.

**Comment by thegreatgabsby** [comment]
@bsendras awesome mate, that worked for me. cheers!

**Comment by pengcao1** [comment]
@bsendras 
great. It's worked for me. thanks

**Comment by crobinson42** [comment]
For me I had to put in my machines local IP to get it to connect

```
Reactotron
  .configure({ host: '192.168.0.148' })
  .useReactNative()
  .connect
```

**Comment by shrikant-thd** [comment]
For me, I was using 
`"reactotron-react-native": "^1.14.0"`
but my Reactotron version was `2.6.0`

I upgraded 
`"reactotron-react-native": "^2.6.0"` 
but that still didn't work for me. So downgraded it back to `1.14.0` and did the same to Reactotron and it finally worked.

**Comment by rmevans9** [comment]
Hmm... it is concerning that you were having connectivity issues on the later version of `reactotron`. I will do some testing to see if I can figure out what is going wrong.

**Comment by kevinvangelder** [comment]
For anyone having connectivity issues, remember that you'll need to run `adb reverse tcp:9090 tcp:9090` every time you reconnect your physical device or emulator. Most likely everyone above has already done this, but for any new users that come here looking for answers it might be helpful.

**Comment by xstable** [comment]
I got some trouble too with this.
I've tried with IP, which didn't work. 
After doing the `adb reverse tcp:9090 tcp:9090` and removing the 'host' property from config() it works now. 
Thanks @kevinvangelder

**Comment by Fredsters** [comment]
I'm using a real ios device for developing. Also I use [https://expo.io/]. 

How can I do the `adb reverse tcp:9090 tcp:9090` thing. In the command line? 
I get: `error: no devices/emulators found`

**Comment by rmevans9** [comment]
Adb is an android utility so it will not work in your case. I would suggest putting your machines up in the host property in your configure call

**Comment by Fredsters** [comment]
I did use it like this `Reactotron.configure({host: '<my-ip-address>'}).useReactNative()` now.... And it worked!
I am sure I already tried it like this before. I tried it now on a differenct Pc and used the msi installation (not the exe) maybe that made a difference? However it works now. Thank you very much.

**Comment by Fredsters** [comment]
Still doesn't work on my other Pc. Don't know why...

**Comment by Fredsters** [comment]
I added a rule in my firewall to open port 9090, now it works on this pc as well. Thanks for help and I hope I can help others in the future.

**Comment by catnuxer** [comment]
Here my setup

Phone connected over wifi -> IP 192.168.0.4
MacbookPro (as server) connected over wifi -> IP 192.168.0.3
Reactotron run in macbook port 9090
React Native run on server port 9091

So first time we should reverse the adb port 9090 to 9091. Which is the reactotron on the phone is looking for port 9090 and the server run on port 9091.

Command step 1 `adb reverse tcp:9090 tcp:9091`

After it we want to connect our phone over wifi, so just run
Command step 2 `adb tcpip 5555`
Command step 3 `adb connect 192.168.0.4:5555`

To run react native I recommend to use script in package.json

Command step 4 `react-native run-android --port 9091`

The last one check your reactotron configuration
```
Reactotron
.configure({
    host: '192.168.0.3',
    name: 'appname'
})
.useReactNative()
.connect();
```

And done reactotron connected.
For troubleshooting make sure adb kill-server and run adb start-server again and check your firewall on your server.

**Comment by gsklee** [comment]
@rmevans9 I think there are two ways to solve this issue:

- We settle on the fact that this is not a Reactotron-specific problem but something endemic to how an Android RN app works with a dev server, create a section in the README and ask everyone to modify their `npm start` script to run `adb reverse` before running `react-native start`

- We could try executing the `adb reverse` shell script from Node.js whenever a new reactotron-core-server instance is being spinning up

**Comment by giscafer** [comment]
same issue

**Comment by rmevans9** [comment]
@gsklee There is actually some code that tries to solve this issue and the fix might be much simpler then that. We have a library that tries to get the IP address from RN directly when using a dev version of the app. I think the problem is no one really knows this and overwrites the `host` option when using configure. The fix might be as simple as "Don't pass host when calling configure". That said @giscafer can you please show us your `configure` call for `reactotron`?

Ref: https://github.com/infinitered/reactotron-react-native/blob/master/src/reactotron-react-native.ts#L21

**Comment by gsklee** [comment]
@rmevans9 Not true; this issue happens when you don't set any option within `.configure()` as well. According to the RN docs the `adb reverse` step is necessary for Android and either the users have to do it themselves or Reactotron can do it for the users; for example this is how React Native Debugger does it:
https://github.com/jhen0409/react-native-debugger/blob/04db26a283532966a55ddb9ecdd7dd3615b1e6d1/app/utils/adb.js

**Comment by rmevans9** [comment]
Hmm then the `getHost` code is not working like it used to. Could probably pull that out while making the change to handle this better.

**Comment by MicxSamonte** [comment]
this issue still persist on my workspace. already did the ff:

- adb reverse the port 9090 (and checked via reverse list)
- set the machine ip on the Reactotron configure host param
- verified if the host machine is visible in the LAN from the device (via pingtools)
- restart packagers and rerun app everytime

I'm using Reactotron 2.15.1, and my RN package versions are the ff:
```
    "reactotron-react-native": "^2.1.0",
    "reactotron-redux": "^2.1.0",
    "reactotron-redux-saga": "^3.0.0",
```

**Comment by ChenhuaWANG22** [comment]
In my case, it's the WIFI NETWORK TYPE which blocks Reactotron to discover my PC.

I use a Windows 10 PC and a real android mobile phone to do testing and they are in the same home network domain. Both mobile phone and PC connect to the router by Wifi.
I have to set the wifi type to PRIVATE NETWORK on PC instead of PUBLIC NETWORK to make Reactotron connect correctly.
![image](https://user-images.githubusercontent.com/17723793/62005922-aab3b880-b16c-11e9-8615-22e4bb902ed6.png)

I discover this during adding a rule in my firewall to open port 9090. 
At the end I found it's not the rules which work but the type of network.
Well, still thank you @Fredsters.

Hope it will help others.

**Comment by dioptre** [comment]
Has anyone seen the issue where you can connect OK, but nothing happens from redux? I do the deferred connect methods and see sagas, and the connect, but seems as if the reactotronRedux is doing nothing? No errors or logs or anything I can see...

**Comment by ameernormie** [comment]
Instead of manually putting in the IP address everytime. It's better if you run the following command if it doesn't connect while on device.
`adb reverse tcp:9090 tcp:9090`

**Comment by pauloemp** [comment]
@ChenhuaWANG22 That's odd but for me the solution was the opposit... I'm working with a phisycal device and trying to connect to Reactotron through LAN. Although, when my network was set as private nothing was happening. I had tried on other networks and it was OK so the problem shouldn't be the code. Only by changing the Network type to public I got it to work again. I don't understand much about this network thing but for anyone having this issue, you should try at least changing the network type.

**Comment by ChenhuaWANG22** [comment]
> @ChenhuaWANG22 That's odd but for me the solution was the opposit... I'm working with a physical device and trying to connect to Reactotron through LAN. Although, when my network was set as private nothing was happening. I had tried on other networks and it was OK so the problem shouldn't be the code. Only by changing the Network type to public I got it to work again. I don't understand much about this network thing but for anyone having this issue, you should try at least changing the network type.

@pauloemp interesting. Consider your discovery, I guess it's something like network security settings blocking Reactotron, which can be set differently for each network type.

**Comment by EhsanSarshar** [comment]
adb reverse tcp:<port_reactotorn_listening> tcp:<port_reactotorn_listening>

**Comment by spyland8k** [comment]
If you dont know which port using from emulator, you can check it after the **app + emulator started**. Follow below these steps;
- Open "cmd" and "adb reverse --list", find the android emulator port. You can see on the line: "host-x tcp:19000 tcp:19000"
- Open the "package.json" file in your project folder.

`"adb": "adb reverse tcp:<reactotronport> tcp:<emulatorport>"` 
- Change the related line `**"adb": "adb reverse tcp:9090 tcp:19000"**` 

- After complete above steps successfully, If you're using **Expo** and **Yarn** Open "**cmd**" in your project.
- "yarn run adb" and "yarn run start", it will be call scripts in your package.json. or you can create custom scipt only one.

Hope that helps.

**Comment by Dax911** [comment]
Same issue different fix... [fedora 34]  this was incredibly painful but I now see why it is so useful.  Never again.

**Comment by AllanChain** [comment]
In my case port 9090 is taken by another program and reactotron is not warning me about that. Switching to an unoccupied port fixes the problem.

**Comment by trickeyd** [comment]
This lib to grab hose ip worked like a charm for me:

https://www.npmjs.com/package/react-native-localhost

As you can't really hardcode an IP in a shared codebase, and adding it to a .env or something feels clumsy

**Comment by joshuayoes** [comment]
This has been fixed in `reactotron-react-native@5.0.4-beta.15`, update your app to this package or above and remove `host` from your Reactotron.configure call if you have it set

<img width="358" alt="Screenshot 2023-11-30 at 1 56 59 PM" src="https://github.com/infinitered/reactotron/assets/37849890/14bba063-d57e-423a-a8e9-8223e26341df">
